cmd/kar-controllers/app/server.go

```diff
@@ -31,59 +16,85 @@ limitations under the License.
 package app
 
 import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"golang.org/x/sync/errgroup"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
-	"net/http"
 
+	"github.com/project-codeflare/multi-cluster-app-dispatcher/cmd/kar-controllers/app/metrics"
 	"github.com/project-codeflare/multi-cluster-app-dispatcher/cmd/kar-controllers/app/options"
 	"github.com/project-codeflare/multi-cluster-app-dispatcher/pkg/controller/queuejob"
 	"github.com/project-codeflare/multi-cluster-app-dispatcher/pkg/health"
 
 	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
 )
 
+
 func buildConfig(master, kubeconfig string) (*rest.Config, error) {
 	if master != "" || kubeconfig != "" {
 		return clientcmd.BuildConfigFromFlags(master, kubeconfig)
 	}
 	return rest.InClusterConfig()
 }
 
-func Run(opt *options.ServerOption) error {
+func Run(ctx context.Context, opt *options.ServerOption) error {
 	config, err := buildConfig(opt.Master, opt.Kubeconfig)
 	if err != nil {
 		return err
 	}
 
-	neverStop := make(chan struct{})
-
 	config.QPS = 100.0
 	config.Burst = 200.0
 
+	g, gCtx := errgroup.WithContext(ctx)
```

Comment:
I think `defer gCtx.cancel()` missing here, which would make sure the context passed to `jobctrl.Run` is cancelled when an error occurs afterwards.